### PR TITLE
[SPARK-46043][SQL] Support create table using DSv2 sources

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR supports `CREATE TABLE ... USING source` for DSv2 sources.

### Why are the changes needed?

To support creating DSv2 tables in SQL. Currently the table create can work but when you select a dsv2 table created in SQL, it fails with this error:
```
org.apache.spark.sql.AnalysisException: org.apache.spark.sql.connector.SimpleDataSourceV2 is not a valid Spark SQL Data Source.
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 1]
```diff
@@ -633,6 +633,38 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+          checkAnswer(
+            sql(s"SELECT * FROM test WHERE i < 3"),
+            Seq(Row(0, 0), Row(1, -1), Row(2, -2)))
+        }
+      }
+    }
+    withTable("test") {
+      val cls = classOf[SchemaRequiredDataSource]
+      withClue(cls.getName) {
+        sql(s"CREATE TABLE test USING ${cls.getName}")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:651]: how does this work? empty table schema?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 2]
```diff
@@ -633,6 +633,38 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
       }
     }
   }
+
+  test("SPARK-46043: Support create table using DSv2 sources") {
+    Seq(classOf[SimpleDataSourceV2], classOf[JavaSimpleDataSourceV2]).foreach { cls =>
+      withClue(cls.getName) {
+        withTable("test") {
+          sql(s"CREATE TABLE test USING ${cls.getName}")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:641]: can we also test what happens if we create table with a schema that does not match the data source?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 3]
```diff
@@ -633,6 +633,38 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+        sql(s"CREATE TABLE test USING ${cls.getName}")
+        checkAnswer(sql(s"SELECT * FROM test"), Nil)
+      }
+    }
+    withTable("test") {
+      val cls = classOf[SupportsExternalMetadataWritableDataSource]
+      withClue(cls.getName) {
+        withTempDir { dir =>
+          sql(s"CREATE TABLE test USING ${cls.getName} OPTIONS (path '${dir.getCanonicalPath}')")
+          checkAnswer(sql(s"SELECT * FROM test"), Nil)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:660]: we should at least test the table schema if there is no data.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala [hunk 4]
```diff
@@ -151,6 +153,27 @@ private[sql] object DataSourceV2Utils extends Logging {
...
+   * Returns the table provider for the given format, or None if it cannot be found.
+   */
+  def getTableProvider(provider: String, conf: SQLConf): Option[TableProvider] = {
+    // Return earlier since `lookupDataSourceV2` may fail to resolve provider "hive" to
+    // `HiveFileFormat`, when running tests in sql/core.
+    if (DDLUtils.isHiveTable(Some(provider))) return None
+    DataSource.lookupDataSourceV2(provider, conf) match {
+      // TODO(SPARK-28396): Currently file source v2 can't work with tables.
+      case Some(_: FileDataSourceV2) => None
+      case o => o
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala:166]:
```
case p: Some(v) if !v.isInstanceOf[FileDataSourceV2] => p
case _ => None
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala [hunk 5]
```diff
@@ -151,6 +153,27 @@ private[sql] object DataSourceV2Utils extends Logging {
...
+    }
+  }
+
+  /**
+   * Check if the provider is a v2 provider.
+   */
+  def isV2Provider(provider: String, conf: SQLConf): Boolean = {
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala:173]: Because `isV2Provider` is only used for  `ResolveSessionCatalog`, move this back to `ResolveSessionCatalog`.

Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala:173]: Good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 6]
```diff
@@ -115,8 +163,40 @@ class V2SessionCatalog(catalog: SessionCatalog)
...
       properties: util.Map[String, String]): Table = {
     import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.TransformHelper
-    val (partitionColumns, maybeBucketSpec, maybeClusterBySpec) = partitions.toSeq.convertTransforms
     val provider = properties.getOrDefault(TableCatalog.PROP_PROVIDER, conf.defaultDataSourceName)
+
+    val (newSchema, newPartitions) = DataSourceV2Utils.getTableProvider(provider, conf) match {
+      case Some(_: SupportsCatalogOptions) =>
+        throw new SparkUnsupportedOperationException(
+          errorClass = "CANNOT_CREATE_DATA_SOURCE_V2_TABLE.CATALOG_OPTIONS_UNSUPPORTED",
+          messageParameters = Map("provider" -> provider))
```
[the remark this answers is not in the document]
Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:172]: @cloud-fan Actually no. CatalogManager constructor takes in a v2SessionCatalog, and here we can't pass in the catalog manager to the constructor of v2 session catalog (circular dependency):
https://github.com/apache/spark/blob/7a0d0411aa02e7e1b6beb393966ada2c54c09870/sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala#L174-L176

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala [hunk 7]
```diff
@@ -31,6 +31,7 @@ trait SimpleTableProvider extends TableProvider {
   def getTable(options: CaseInsensitiveStringMap): Table
 
   private[this] var loadedTable: Table = _
+
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala:34]: Please revert this line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 8]
```diff
@@ -73,7 +74,44 @@ class V2SessionCatalog(catalog: SessionCatalog)
...
-      V1Table(catalog.getTableMetadata(ident.asTableIdentifier))
+      val table = catalog.getTableMetadata(ident.asTableIdentifier)
+      if (table.provider.isDefined) {
+        DataSourceV2Utils.getTableProvider(table.provider.get, conf) match {
+          case Some(provider) =>
+            // Get the table properties during creation and append the path option
+            // to the properties.
+            val tableProperties = table.properties
+            val pathOption = table.storage.locationUri.map("path" -> CatalogUtils.URIToString(_))
+            val properties = tableProperties ++ pathOption
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:85]:
```suggestion
            val properties = table.properties ++
              table.storage.locationUri.map("path" -> CatalogUtils.URIToString(_))
```

**Review by beliefer — COMMENTED** [review]
Looks good!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala [hunk 9]
```diff
@@ -45,7 +45,6 @@ trait SimpleTableProvider extends TableProvider {
       schema: StructType,
       partitioning: Array[Transform],
       properties: util.Map[String, String]): Table = {
-    assert(partitioning.isEmpty)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala:48]: Do you know the reason about the assert? If not, please revert it.

**Review by beliefer — APPROVED** [review]
LGTM. cc @cloud-fan @huaxingao

File: common/utils/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -149,6 +149,19 @@
     ],
     "sqlState" : "42846"
   },
+  "CANNOT_CREATE_DATA_SOURCE_V2_TABLE" : {
+    "message" : [
+      "Failed to create data source V2 table:"
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:154]: shall we include the table name?

File: common/utils/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -149,6 +149,19 @@
     ],
     "sqlState" : "42846"
   },
+  "CANNOT_CREATE_DATA_SOURCE_V2_TABLE" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:152]: I can't find other errors that mention data source v2. I think it's a developer thing and we should not expose it to end users via error message. How about just `CANNOT_CREATE_TABLE`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 12]
```diff
@@ -61,7 +61,14 @@ case class DataSourceV2Relation(
       Nil
   }
 
-  override def name: String = table.name()
+  override def name: String = {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+    (catalog, identifier) match {
+      case (Some(cat), Some(ident)) => s"${quoteIdentifier(cat.name())}.${ident.quoted}"
+      case (None, Some(ident)) => ident.quoted
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala:68]: I don't think this can happen. We can add an assert.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala (hunk 9 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SimpleTableProvider.scala:48]: +1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 13]
```diff
@@ -115,9 +164,44 @@ class V2SessionCatalog(catalog: SessionCatalog)
...
+          // Infer the schema and partitions and store them in the catalog.
+          (tableProvider.inferSchema(dsOptions), tableProvider.inferPartitioning(dsOptions))
+        } else if (partitions.isEmpty) {
+          (schema, tableProvider.inferPartitioning(dsOptions))
+        } else {
+          (schema, partitions)
+        }
+
+      case _ =>
+        (schema, partitions)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:200]: shall we fail here if it's not a valid data source?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/SupportsCatalogOptionsSuite.scala [hunk 14]
```diff
@@ -25,7 +25,7 @@ import scala.util.Try
 
 import org.scalatest.BeforeAndAfter
 
-import org.apache.spark.SparkException
+import org.apache.spark.{SparkException, SparkUnsupportedOperationException}
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/SupportsCatalogOptionsSuite.scala:28]: unnecessary change?

**Comment by allisonwang-db** [comment]
The test failure seems unrelated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 15]
```diff
@@ -114,10 +154,46 @@ class V2SessionCatalog(catalog: SessionCatalog)
...
+        }
+        // V2CreateTablePlan does not allow non-empty partitions when schema is empty. This
+        // is checked in `PreProcessTableCreation` rule.
+        assert(partitions.isEmpty,
+          s"Partitions should be empty when the schema is empty: ${partitions.mkString(", ")}")
+        (schema, partitions)
+
+      case Some(tableProvider) =>
+        assert(tableProvider.supportsExternalMetadata())
+        lazy val dsOptions = new CaseInsensitiveStringMap(properties)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:178]: do we need to put the path option?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:178]: I think we can add a new method to create ds options from a `CatalogTable`, to save duplicated code.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 16]
```diff
@@ -3366,8 +3367,21 @@ class DataSourceV2SQLSuiteV2Filter extends DataSourceV2SQLSuite {
 
 /** Used as a V2 DataSource for V2SessionCatalog DDL */
 class FakeV2Provider extends SimpleTableProvider {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:3369]: Can we avoid extending `SimpleTableProvider` here? I think it's not meant to support external metadata.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 17]
```diff
@@ -633,6 +634,95 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+            sql(s"SELECT * FROM test WHERE i < 3"),
+            Seq(Row(0, 0), Row(1, -1), Row(2, -2)))
+        }
+        // Create a table with non-empty schema is not allowed.
+        checkError(
+          exception = intercept[SparkUnsupportedOperationException] {
+            sql(s"CREATE TABLE test(a INT, b INT) USING ${cls.getName}")
+          },
+          errorClass = "CANNOT_CREATE_DATA_SOURCE_TABLE.EXTERNAL_METADATA_UNSUPPORTED",
+          parameters = Map("tableName" -> "default.test", "provider" -> cls.getName)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:654]: it should be `"`default`.`test`"`

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 18]
```diff
@@ -633,6 +634,95 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+        )
+      }
+    }
+  }
+
+  test("SPARK-46043: create table in SQL with schema required data source") {
+    val cls = classOf[SchemaRequiredDataSource]
+    val e = intercept[IllegalArgumentException] {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:662]: oh, it doesn't have an error class?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 19]
```diff
@@ -633,6 +634,95 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+        sql("SELECT * FROM test").collect()
+      }
+      assert(e.getMessage.contains(
+        "java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2"))
+    }
+  }
+
+  test("SPARK-46043: create table in SQL with partitioning required data source") {
+    val cls = classOf[PartitionsRequiredDataSource]
+    val e = intercept[IllegalArgumentException](
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:690]: oh it's thrown directly from the data source?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 20]
```diff
@@ -61,7 +61,19 @@ case class DataSourceV2Relation(
...
   }
 
-  override def name: String = table.name()
+  override def name: String = {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+    (catalog, identifier) match {
+      case (Some(cat), Some(ident)) => s"${quoteIfNeeded(cat.name())}.${ident.quoted}"
+      case (None, None) => table.name()
+      case _ =>
+        throw new IllegalArgumentException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala:70]: this should be `SparkException.internalError`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
